test/server/charset-test.js

```diff
@@ -0,0 +1,329 @@
+                charsetUrl,
+                iconv.encode(processedScript, charsetUrl).toString(),
+                done
+            );
+        });
+    });
+
+    describe('Other resources', function () {
+        var testResourceCharset = function (expectedBody, charsetStr, url) {
+            return new Promise(function (resolve) {
+                request(url, function (err, res, body) {
+                    compareCode(body, iconv.encode(expectedBody, charsetStr, { addBOM: /\/bom$/.test(url) }).toString());
+
+                    resolve();
+                });
+            });
+        };
+
+        it('Should set content charset for manifest', function (done) {
+            var processedManifest = manifestProcessor.processResource(manifestSrc, null, null, getProxyUrl);
+            var resourceUrl       = 'http://127.0.0.1:2000/other-resource/manifest/';
+
+            Promise
+                .all([
+                    testResourceCharset(processedManifest, charsetBOM, getProxyUrl(resourceUrl + 'bom')),
+                    testResourceCharset(processedManifest, charsetContentType, getProxyUrl(resourceUrl + 'content-type')),
+                    testResourceCharset(processedManifest, charsetDefault, getProxyUrl(resourceUrl + 'default'))
+                ])
+                .then(function () {
+                    done();
+                })
+                .catch(done);
+        });
+
+        it('Should set content charset for stylesheet', function (done) {
+            var processedStylesheet = stylesheetProcessor.processResource(stylesheetSrc, null, null, getProxyUrl);
+            var resourceUrl         = 'http://127.0.0.1:2000/other-resource/stylesheet/';
+
+            Promise
+                .all([
+                    testResourceCharset(processedStylesheet, charsetBOM, getProxyUrl(resourceUrl + 'bom')),
+                    testResourceCharset(processedStylesheet, charsetContentType, getProxyUrl(resourceUrl + 'content-type')),
+                    testResourceCharset(processedStylesheet, charsetDefault, getProxyUrl(resourceUrl + 'default'))
+                ])
+                .then(function () {
+                    done();
+                })
+                .catch(done);
+        });
+    });
+
+    it('Should correctly determine the charset from BOM', function (done) {
+        var testBOM = function (bomCharset, contentTypeHeader) {
+            var inputData = iconv.encode(pageWithoutMetaSrc, bomCharset, { addBOM: true });
+            var charset   = new Charset();
+
+            charset.fromContentType(contentTypeHeader);
+
+            return contentUtils.decodeContent(inputData, '', charset).then(function (decoded) {
+                expect(decoded).eql(pageWithoutMetaSrc);
+                expect(charset.get()).eql(bomCharset);
+                expect(charset.isFromBOM()).to.be.true;
+
+                return contentUtils.encodeContent(pageWithoutMetaSrc, '', charset).then(function (encoded) {
```

Comment:
It can be refactored to avoid additional nesting level:
```js
  return contentUtils.decodeContent(inputData, '', charset)
     .then(function(decoded) {
         expect(decoded).eql(pageWithoutMetaSrc);
         expect(charset.get()).eql(bomCharset);
         expect(charset.isFromBOM()).to.be.true;

         return contentUtils.encodeContent(pageWithoutMetaSrc, '', charset);
     })
     .then(function(encoded) {
         expect(encoded).to.deep.equal(inputData);
     });
```